### PR TITLE
Add PercentileKLL aggregation function

### PR DESCRIPTION
**Pull request by cbalci**
Introducing a new approximate percentile calculation function `PercentileKLL` and its variations (MV & Raw), using Apache Datasketches libraries 'KLL'.

This is part of a proposal to improve Apache Datasketches support in Pinot: 
[(Google Docs Link) [Proposal] Improved Apache DataSketches Support in Pinot](https://docs.google.com/document/d/1ctmKVRi67lpO6x1RYKDvDYf05EZx2Vbs2OnUudYP-bU/edit )

Some advantages listed and discussed in the linked document:
- [Well defined](https://datasketches.apache.org/docs/KLL/KLLAccuracyAndSize.html) error bound ([comparison](https://datasketches.apache.org/docs/QuantilesStudies/KllSketchVsTDigest.html) to t-Digest)
- Faster updates, serialization/deserialization
- Binary compatibility with external systems, hence the ability to use Pinot as sketch store
- Ability to compute ‘Rank’ and ‘Histogram’ besides ‘Percentile’
- Feature parity with Druid

Please leave design related comments on the linked document and code related comments in this PR.

**Testing**
- Added unit tests to cover basic use cases that call `PercentileKLL`, `PercentileKLLMV`,  `PercentileRawKLL`, `PercentileRawKLLMV`
- Added tests to cover group by scenarios
- Manually tested ingesting raw (externally generated) data sketches

`feature` `performance`

### CODE REVIEW

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/customobject/SerializedKLL.java [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+  private final KllDoublesSketch _sketch;
+
+  public SerializedKLL(KllDoublesSketch sketch, double percentile) {
+    _sketch = sketch;
+    _quantile = percentile / 100.0;
+  }
+
+  @Override
+  public int compareTo(SerializedKLL other) {
+    checkArgument(other._quantile == _quantile, "Quantile numbers don't matchå");
```
Comment (chiquelo) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/customobject/SerializedKLL.java:41]: Is this character intended?

File: pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/AggregationFunctionType.java [hunk 2]
```diff
@@ -142,6 +146,10 @@ public static AggregationFunctionType getAggregationFunctionType(String function
         return PERCENTILETDIGEST;
       } else if (remainingFunctionName.equals("RAWTDIGEST") || remainingFunctionName.matches("RAWTDIGEST\\d+")) {
         return PERCENTILERAWTDIGEST;
+      } else if (remainingFunctionName.equals("KLL") || remainingFunctionName.matches("KLL\\d+")) {
+        return PERCENTILETDIGEST;
```
Comment (chiquelo) [inline pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/AggregationFunctionType.java:150]: This seems like an error

**Review by chiquelo — CHANGES_REQUESTED** [review]

File: pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/AggregationFunctionFactory.java [hunk 3]
```diff
@@ -77,6 +88,14 @@ public static AggregationFunction getAggregationFunction(FunctionContext functio
             // PercentileRawTDigest
             String percentileString = remainingFunctionName.substring(10);
             return new PercentileRawTDigestAggregationFunction(firstArgument, parsePercentileToInt(percentileString));
+          } else if (remainingFunctionName.matches("KLL\\d+")) {
+            // PercentileKLL
+            String percentileString = remainingFunctionName.substring(3);
```
Comment (chenboat) [inline pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/AggregationFunctionFactory.java:93]: why 3? can you add an comment?

File: pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/AggregationFunctionFactory.java [hunk 4]
```diff
@@ -77,6 +88,14 @@ public static AggregationFunction getAggregationFunction(FunctionContext functio
...
+            return new PercentileKLLAggregationFunction(firstArgument, parsePercentileToInt(percentileString));
+          } else if (remainingFunctionName.matches("RAWKLL\\d+")) {
+            // PercentileRawKLL
+            String percentileString = remainingFunctionName.substring(6);
```
Comment (chenboat) [inline pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/AggregationFunctionFactory.java:97]: Add comment and/or example to explain why 6? Ditto to all constant numbers used below.

File: pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/PercentileKLLAggregationFunction.java [hunk 5]
```diff
@@ -0,0 +1,248 @@
...
+  }
+
+  @Override
+  public void aggregate(int length, AggregationResultHolder aggregationResultHolder,
+      Map<ExpressionContext, BlockValSet> blockValSetMap) {
+    BlockValSet valueSet = blockValSetMap.get(_expression);
+    DataType valueType = valueSet.getValueType();
+    KllDoublesSketch sketch = getOrCreateSketch(aggregationResultHolder);
+
+    if (valueType == DataType.BYTES) {
```
Comment (chenboat) [inline pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/PercentileKLLAggregationFunction.java:105]: javadoc to explain the difference caused by valueType?

File: pinot-core/src/test/java/org/apache/pinot/queries/PercentileKLLQueriesTest.java [hunk 6]
```diff
@@ -0,0 +1,256 @@
...
+
+import static org.testng.Assert.assertEquals;
+import static org.testng.Assert.assertNotNull;
+
+
+/**
+ * Tests for PERCENTILE_KLL aggregation function.
+ *
+ * <ul>
+ *   <li>Generates a segment with a double column: a KLL column and a group-by column</li>
```
Comment (chenboat) [inline pinot-core/src/test/java/org/apache/pinot/queries/PercentileKLLQueriesTest.java:66]: Are you talking about three columns here? If so, change : to ,

File: pinot-core/src/main/java/org/apache/pinot/core/common/ObjectSerDeUtils.java [hunk 7]
```diff
@@ -918,6 +923,26 @@ public Sketch deserialize(ByteBuffer byteBuffer) {
     }
   };
 
+  public static final ObjectSerDe<KllSketch> KLL_SKETCH_SER_DE = new ObjectSerDe<KllSketch>() {
```
[the remark this answers is not in the document]
Comment (cbalci) [reply pinot-core/src/main/java/org/apache/pinot/core/common/ObjectSerDeUtils.java]: Good point, updated.

File: pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/AggregationFunctionFactory.java (hunk 3 above)
Comment (cbalci) [reply pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/AggregationFunctionFactory.java:93]: 3 is the number of characters already consumed ("KLL"). All the other if branches uses the same logic so I didn't think it requires extra explanation.

File: pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/AggregationFunctionFactory.java (hunk 4 above)
Comment (cbalci) [reply pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/AggregationFunctionFactory.java:97]: Same as above. "RAWKLL" has six characters, so we skip ahead to capture the rest of the function name.

File: pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/PercentileKLLAggregationFunction.java [hunk 8]
```diff
@@ -0,0 +1,248 @@
...
+  public PercentileKLLAggregationFunction(List<ExpressionContext> arguments) {
+    super(arguments.get(0));
+
+    // Check that there are correct number of arguments
+    int numArguments = arguments.size();
+    Preconditions.checkArgument(numArguments == 2 || numArguments == 3,
+        "Expecting 2 or 3 arguments for PercentileKLL function: "
+            + "PERCENTILE_KLL(column, percentile, k=200");
+
+    _percentile = arguments.get(1).getLiteral().getDoubleValue();
```
[the remark this answers is not in the document]
Comment (cbalci) [reply pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/PercentileKLLAggregationFunction.java:77]: Good point, added a check.

File: pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/PercentileKLLAggregationFunction.java (hunk 5 above)
Comment (cbalci) [reply pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/PercentileKLLAggregationFunction.java:112]: Good suggestion. Updated the comments below and added explanation to function javadoc.

File: pinot-core/src/test/java/org/apache/pinot/queries/PercentileKLLQueriesTest.java (hunk 6 above)
Comment (cbalci) [reply pinot-core/src/test/java/org/apache/pinot/queries/PercentileKLLQueriesTest.java]: Fixed.

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/customobject/SerializedKLL.java (hunk 1 above)
Comment (cbalci) [reply pinot-segment-local/src/main/java/org/apache/pinot/segment/local/customobject/SerializedKLL.java]: Not really, fixed.

File: pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/AggregationFunctionType.java (hunk 2 above)
Comment (cbalci) [reply pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/AggregationFunctionType.java]: Great catch! Fixed.

File: pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/PercentileKLLAggregationFunction.java [hunk 9]
```diff
@@ -0,0 +1,248 @@
...
+    if (numArguments == 3) {
+      _kValue = arguments.get(2).getLiteral().getIntValue();
```
[the remark this answers is not in the document]
Comment (cbalci) [reply pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/PercentileKLLAggregationFunction.java:81]: The library already returns a good exception (`SketchesArgumentException: K must be >= 8 and <= 65535`), so I didn't see much value in doing another check here.

**Review by chiquelo — APPROVED** [review]
lgtm

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/customobject/SerializedKLL.java [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+import org.apache.datasketches.kll.KllDoublesSketch;
+
+import static com.google.common.base.Preconditions.checkArgument;
+
+/**
+ * Serialized and comparable version of KllDoublesSketch.
+ * Compares two sketches for a specific percentile value.
+ */
+public class SerializedKLL implements Comparable<SerializedKLL> {
+  private final double _quantile;
```
Comment (nklijia2011) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/customobject/SerializedKLL.java:31]: why do we need _quantile inside this SerializedKLL? 
if we want to return the KllDoublesSketch as byte[], the toString() is for that purpose?

Comment (cbalci) [reply pinot-segment-local/src/main/java/org/apache/pinot/segment/local/customobject/SerializedKLL.java:31]: Good question. That is used when we want to sort (ORDER BY) sketches.

**Comment by navina** [comment]
@cbalci / @chenboat :  has this feature been documented ? is it available in the OSS docs?

**Comment by cbalci** [comment]
@navina, not yet. I'll add it to OSS documentation soon.

**Pull request merged**